src/treelearner/feature_histogram.hpp

```diff
@@ -557,23 +569,25 @@ class FeatureHistogram {
         if (sum_left_hessian < meta_->config->min_sum_hessian_in_leaf) break;
 
         double sum_left_gradient = sum_gradient - sum_right_gradient;
-        // current split gain
-        double current_gain = GetSplitGains(sum_left_gradient, sum_left_hessian, sum_right_gradient, sum_right_hessian,
-          meta_->config->lambda_l1, meta_->config->lambda_l2, meta_->config->max_delta_step,
-          min_constraint, max_constraint, meta_->monotone_type);
-        // gain with split is worse than without split
-        if (current_gain <= min_gain_shift) continue;
+        if (!meta_->config->extra_trees || t - 1 + offset == rand_threshold) {
```

Comment:
maybe change `FindBestThresholdSequence` to the template method: `template<bool is_rand> FindBestThresholdSequence` ?

Reply:
with the template, the function will be expanded at compiling, and don't affect the run-time performance.